src/Access/Authentication.h

```diff
@@ -90,6 +96,7 @@ class Authentication
     /// Checks if the provided password is correct. Returns false if not.
     /// User name and external authenticators' info are used only by some specific authentication type (e.g., LDAP_SERVER).
     bool isCorrectPassword(const String & password_, const String & user_, const ExternalAuthenticators & external_authenticators) const;
+    bool isCorrectPasswordLDAP(const String & password_, const String & user_, const ExternalAuthenticators & external_authenticators, const LDAPSearchParamsList * search_params = nullptr, LDAPSearchResultsList * search_results = nullptr) const;
```

Comment:
It doesn't look good to me. We had a single function `isCorrectPassword()` and now we're going to have an extra function just for one authentication type. It seems you could easily avoid that because there is one place where those extra arguments are passed from - it's `LDAPAccessStorage`, and `LDAPAccessStorage` always uses only the same authentication type and `LDAPAccessStorage` could call `ldap_client.authenticate()` directly without spoiling the general interface.

Reply:
I didn't want to implement the authentication logic in different places. That would make the code misleading and hard to maintain.

Reply:
By the way, the idea of `Credential` would solve this particular problem quite nicely: the authentication logic would be encapsulated in each `Credential` implementation, and accessed by a common interface inside `Authentication` class calls.

Reply:
> LDAPAccessStorage could call ldap_client.authenticate() directly without spoiling the general interface.

Once #15988 is merged, I'll be able to call `ExternalAuthenticators::checkLDAPCredentials(...)` from `LDAPAccessStorage` directly...

Reply:
https://github.com/ClickHouse/ClickHouse/pull/15988 merged

Reply:
Changed.